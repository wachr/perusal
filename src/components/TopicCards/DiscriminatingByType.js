export const TOPIC_ARRAY = Symbol("TOPIC_ARRAY");
export const TOPIC_OBJECT = Symbol("TOPIC_OBJECT");
export const TOPIC_STRING = Symbol("TOPIC_STRING");
export const TOPIC_UNKNOWN = Symbol("TOPIC_UNKNOWN");
export function discriminating(topic) {
  return new DiscriminatingByType(topic);
}

class DiscriminatingByType {
  #topic;
  #topicTypes = [];
  #cases = [];
  #defaultCase = null;

  static discriminator(topic) {
    if (Array.isArray(topic)) return TOPIC_ARRAY;
    if (topic && topic.topicTitle) return TOPIC_OBJECT;
    if (topic && typeof topic === "string") return TOPIC_STRING;
    return TOPIC_UNKNOWN;
  }

  constructor(topic) {
    this.#topic = topic;
  }

  considering(...topicTypes) {
    this.#topicTypes = topicTypes;
    return this;
  }

  renderWith(renderFn) {
    this.#cases.push([this.#topicTypes, renderFn]);
    return this;
  }

  defaultTo(renderFn) {
    this.#defaultCase = renderFn;
    return this;
  }

  type() {
    return DiscriminatingByType.discriminator(this.#topic);
  }

  render() {
    return (
      this.#cases.reduce((rendered, [topicTypes, renderFn]) => {
        if (rendered) return rendered;
        if (topicTypes.includes(this.type())) return renderFn();
        return null;
      }, null) ||
      (this.#defaultCase && this.#defaultCase()) ||
      null
    );
  }
}
