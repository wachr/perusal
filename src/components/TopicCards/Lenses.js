export function selectByPath(topic, path) {
  return editByPath(topic, path, subtopic => subtopic);
}

export function editByPath(topic, path, calculateNewValue) {
  if (!Array.isArray(path) || path.length === 0)
    return calculateNewValue(topic);
  return (
    path.reduce((currentTopic, pathElement) => {
      if (!pathElement) return calculateNewValue(currentTopic);
      if (
        pathElement === currentTopic ||
        pathElement === currentTopic.topicTitle
      )
        return calculateNewValue(currentTopic);
      const subtopic =
        currentTopic?.topicSubtopics &&
        currentTopic.topicSubtopics.find(
          subtopic =>
            subtopic?.topicTitle === pathElement || subtopic === pathElement
        );
      return subtopic && calculateNewValue(subtopic);
    }, topic) || topic
  );
}
