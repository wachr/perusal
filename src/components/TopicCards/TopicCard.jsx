import { h } from "preact";
import Box from "@mui/material/Box";
import Card from "@mui/material/Card";
import Divider from "@mui/material/Divider";
import PropTypes from "prop-types";
import Stack from "@mui/material/Stack";
import Typography from "@mui/material/Typography";

import {
  TOPIC_OBJECT,
  TOPIC_STRING,
  discriminating
} from "./DiscriminatingByType";
import TopicSubtopics from "./TopicSubtopics";
import TopicTitle from "./TopicTitle";
import TopicUnrenderableAlert from "./TopicUnrenderableAlert";

function TopicCard({ topic }) {
  return (
    <Box sx={{ width: "400px", padding: "3px" }}>
      <Card>
        {discriminating(topic)
          .considering(TOPIC_STRING)
          .renderWith(() => <TopicTitle topic={topic} />)
          .considering(TOPIC_OBJECT)
          .renderWith(() => (
            <Stack>
              <TopicTitle topic={topic} />
              {(topic.topicDetails || topic.topicSubtopics) && <Divider />}
              {topic.topicDetails && (
                <Typography variant="body" data-testid="topic-details">
                  {topic.topicDetails}
                </Typography>
              )}
              <TopicSubtopics topic={topic} />
            </Stack>
          ))
          .defaultTo(() => <TopicUnrenderableAlert topic={topic} />)
          .render()}
      </Card>
    </Box>
  );
}

export const SINGLE_TOPIC_PROP_TYPE = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({
    topicTitle: PropTypes.string.isRequired,
    topicDetails: PropTypes.string,
    topicSubtopics: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          topicTitle: PropTypes.string.isRequired
        })
      ]).isRequired
    )
  })
]).isRequired;

TopicCard.proptypes = {
  topic: SINGLE_TOPIC_PROP_TYPE.isRequired
};

export default TopicCard;
