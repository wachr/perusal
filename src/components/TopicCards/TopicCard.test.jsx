import { h } from "preact";
import { fireEvent, render, screen } from "@testing-library/preact";

import { TOPIC_STRING, discriminating } from "./DiscriminatingByType";
import { selectByPath } from "./Lenses";
import TopicCard from "./TopicCard";

jest.mock("wouter-preact", () => {
  const wouterActual = jest.requireActual("wouter-preact");
  return {
    ...wouterActual,
    useRoute: jest.fn().mockReturnValue([false, {}])
  };
});
const useRoute = require("wouter-preact").useRoute;

afterEach(() => {
  useRoute.mockClear();
});

describe(TopicCard.name, () => {
  it("should render with minimal state", () => {
    render(<TopicCard topic={"foo"} />);
    const topicCard = screen.queryByTestId("topic-card");
    expect(topicCard).toMatchSnapshot();
  });

  it("should be able to render a minimal subtopic", () => {
    const topic = {
      topicTitle: "foo",
      topicSubtopics: ["bar"]
    };
    const path = ["foo", "bar"];
    useRoute.mockReturnValueOnce([true, { topicPath: path.join("/") }]);
    render(<TopicCard topic={topic} />);
    expect(selectByPath(topic, path)).toBe("bar");
    expect(discriminating(selectByPath(topic, path)).type()).toBe(TOPIC_STRING);
    const topicCard = screen.queryByTestId("topic-card");
    expect(topicCard).toMatchSnapshot();
    expect(topicCard).toHaveTextContent("bar");
  });

  it("should be able to lens into a subtopic", () => {
    const topic = {
      topicTitle: "foo",
      topicSubtopics: ["bar", "baz"]
    };
    const path = ["foo", "bar"];
    useRoute.mockReturnValueOnce([true, { topicPath: path.join("/") }]);
    render(<TopicCard topic={topic} />);
    const topicCard = screen.queryByTestId("topic-card");
    expect(topicCard).not.toHaveTextContent("foo");
    expect(topicCard).toHaveTextContent("bar");
    expect(topicCard).not.toHaveTextContent("baz");
  });
});
