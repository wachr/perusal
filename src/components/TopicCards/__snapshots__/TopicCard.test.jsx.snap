// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicCard should be able to render a minimal subtopic 1`] = `
<div
  data-testid="topic-card"
  id="TopicCard"
>
  <div
    class="MuiBox-root css-1s8p6ck"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        data-testid="topic-title"
      >
        <a
          class="MuiTypography-root MuiTypography-h6 MuiLink-root MuiLink-underlineAlways css-igcn10-MuiTypography-root-MuiLink-root"
          href="bar"
        >
          bar
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopicCard should render with minimal state 1`] = `
<div
  data-testid="topic-card"
  id="TopicCard"
>
  <div
    class="MuiBox-root css-1s8p6ck"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        data-testid="topic-title"
      >
        <a
          class="MuiTypography-root MuiTypography-h6 MuiLink-root MuiLink-underlineAlways css-igcn10-MuiTypography-root-MuiLink-root"
          href="foo"
        >
          foo
        </a>
      </div>
    </div>
  </div>
</div>
`;
